File: examples/shutdown_client.py
#!/usr/bin/env python3
"""
Example client to demonstrate server shutdown functionality.
"""
import argparse
import sys
from pathlib import Path

import grpc

# Add the proto directory to the path
sys.path.insert(0, str(Path(__file__).parent.parent / "proto"))

# Import the generated gRPC stubs
import photorec_pb2
import photorec_pb2_grpc


def shutdown_server(host="localhost", port=50051, force=False, reason=""):
    """
    Send a shutdown request to the PhotoRec gRPC server.

    Args:
        host: Server host address
        port: Server port
        force: Force shutdown even with active recoveries
        reason: Optional reason for shutdown
    """
    # Create gRPC channel
    channel = grpc.insecure_channel(f"{host}:{port}")

    # Create client stub
    stub = photorec_pb2_grpc.PhotoRecServiceStub(channel)

    try:
        # Create shutdown request
        request = photorec_pb2.ShutdownRequest(force=force, reason=reason)

        print(f"Sending shutdown request to {host}:{port}...")
        print(f"Force: {force}")
        if reason:
            print(f"Reason: {reason}")

        # Send shutdown request
        response = stub.Shutdown(request)

        if response.success:
            print("✓ Shutdown request successful")
            print(f"Message: {response.message}")
        else:
            print("✗ Shutdown request failed")
            print(f"Error: {response.error_message}")
            return False

    except grpc.RpcError as e:
        print(f"✗ gRPC error: {e.code()}: {e.details()}")
        return False
    except Exception as e:
        print(f"✗ Unexpected error: {e}")
        return False
    finally:
        channel.close()

    return True


def main():
    """Main function."""
    parser = argparse.ArgumentParser(
        description="Shutdown PhotoRec gRPC server remotely",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    parser.add_argument("--host", "-H", default="localhost", help="Server host address")

    parser.add_argument("--port", "-p", type=int, default=50051, help="Server port")

    parser.add_argument(
        "--force",
        "-f",
        action="store_true",
        help="Force shutdown even if there are active recovery sessions",
    )

    parser.add_argument(
        "--reason", "-r", default="", help="Optional reason for shutdown"
    )

    args = parser.parse_args()

    # Shutdown the server
    success = shutdown_server(
        host=args.host, port=args.port, force=args.force, reason=args.reason
    )

    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: examples/shutdown_client.cpp
#include <iostream>
#include <memory>
#include <string>
#include <grpcpp/grpcpp.h>
#include "testdisk.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using testdisk::TestDiskService;
using testdisk::ShutdownRequest;
using testdisk::ShutdownResponse;

class ShutdownClient {
public:
    explicit ShutdownClient(std::shared_ptr<Channel> channel)
        : stub_(TestDiskService::NewStub(channel)) {}

    bool ShutdownServer(bool force = false, const std::string& reason = "") {
        ShutdownRequest request;
        request.set_force(force);
        if (!reason.empty()) {
            request.set_reason(reason);
        }

        ShutdownResponse response;
        ClientContext context;

        std::cout << "Sending shutdown request..." << std::endl;
        std::cout << "Force: " << (force ? "true" : "false") << std::endl;
        if (!reason.empty()) {
            std::cout << "Reason: " << reason << std::endl;
        }

        Status status = stub_->Shutdown(&context, request, &response);

        if (status.ok()) {
            if (response.success()) {
                std::cout << "✓ Shutdown request successful" << std::endl;
                std::cout << "Message: " << response.message() << std::endl;
                return true;
            } else {
                std::cout << "✗ Shutdown request failed" << std::endl;
                std::cout << "Error: " << response.error_message() << std::endl;
                return false;
            }
        } else {
            std::cout << "✗ gRPC error: " << status.error_code() << ": " 
                      << status.error_message() << std::endl;
            return false;
        }
    }

private:
    std::unique_ptr<TestDiskService::Stub> stub_;
};

void PrintUsage(const char* program_name) {
    std::cout << "Usage: " << program_name << " [OPTIONS]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  --host, -H HOST      Server host address (default: localhost)" << std::endl;
    std::cout << "  --port, -p PORT      Server port (default: 50051)" << std::endl;
    std::cout << "  --force, -f          Force shutdown even with active recoveries" << std::endl;
    std::cout << "  --reason, -r REASON  Optional reason for shutdown" << std::endl;
    std::cout << "  --help, -h           Show this help message" << std::endl;
    std::cout << std::endl;
    std::cout << "Examples:" << std::endl;
    std::cout << "  " << program_name << " --host localhost --port 50051" << std::endl;
    std::cout << "  " << program_name << " --force --reason \"Maintenance\"" << std::endl;
}

int main(int argc, char* argv[]) {
    std::string host = "localhost";
    int port = 50051;
    bool force = false;
    std::string reason = "";

    // Parse command line arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        
        if (arg == "--help" || arg == "-h") {
            PrintUsage(argv[0]);
            return 0;
        } else if (arg == "--host" || arg == "-H") {
            if (i + 1 < argc) {
                host = argv[++i];
            } else {
                std::cerr << "Error: --host requires an argument" << std::endl;
                return 1;
            }
        } else if (arg == "--port" || arg == "-p") {
            if (i + 1 < argc) {
                port = std::stoi(argv[++i]);
            } else {
                std::cerr << "Error: --port requires an argument" << std::endl;
                return 1;
            }
        } else if (arg == "--force" || arg == "-f") {
            force = true;
        } else if (arg == "--reason" || arg == "-r") {
            if (i + 1 < argc) {
                reason = argv[++i];
            } else {
                std::cerr << "Error: --reason requires an argument" << std::endl;
                return 1;
            }
        } else {
            std::cerr << "Unknown option: " << arg << std::endl;
            std::cerr << "Use --help for usage information" << std::endl;
            return 1;
        }
    }

    try {
        // Create channel
        std::string server_address = host + ":" + std::to_string(port);
        auto channel = grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials());

        // Create client
        ShutdownClient client(channel);

        std::cout << "Connecting to TestDisk gRPC server at " << server_address << std::endl;

        // Send shutdown request
        bool success = client.ShutdownServer(force, reason);

        return success ? 0 : 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}